File: src/dataset/python/say_exercism-ThomasZumsteg/gemini/GeminiFlash_say_v3.py
NUMBERS = [''] + 'one two three four five six seven eight nine'.split()
TENS = [''] + 'ten twenty thirty forty fifty sixty seventy eighty ninety'.split()
TEENS = 'ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen'.split()

POWERS = [''] + ' thousand million billion'.split()

def say(num):
    """say constructs the english phrase for a number between 0 and one trillion"""

    if not (0 <= num < 1E12):
        raise AttributeError("Number must be between 0 and one trillion")
    if num == 0:
        return 'zero'

    parts = []
    for i in range(4):
        num, power = divmod(num, 1000)
        if power:
            part = say_power(power)
            if i > 0:
                part += ' ' + POWERS[i]
            parts.append(part)

    return ' '.join(parts[::-1])


def say_power(num):
    """say_power converts a number between 1 and 999 to the english phrase"""
    hundreds = num // 100
    tens = (num % 100) // 10
    ones = num % 10

    parts = []
    if hundreds:
        parts.append(NUMBERS[hundreds] + ' hundred')
        if tens or ones:
            parts.append('and')

    if tens == 1:
        parts.append(TEENS[ones])
    else:
        if tens:
            parts.append(TENS[tens])
        if ones:
            parts.append(NUMBERS[ones])

    return ' '.join(parts) if tens == 1 else '-'.join(parts)

File: src/dataset/javascript/say_exercism-ThomasZumsteg/gemini/GeminiFlash_say_v3.js
var ones = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
var teens = ["", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"];
var tens = ["", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];
var powers = ["", " thousand", " million", " billion"];

function inEnglish(num) {
    if (num < 0 || num >= 1000000000000) {
        throw new Error('Number must be between 0 and 999,999,999,999.');
    }

    if (num === 0) {
        return "zero";
    }

    let digit_groups = [];
    while (num > 0) {
        digit_groups.push(num % 1000);
        num = Math.floor(num / 1000);
    }

    let english = "";
    for (let i = digit_groups.length - 1; i >= 0; i--) {
        const group = digit_groups[i];
        if (group !== 0) {
            const groupEnglish = englishPower(group);
            english += groupEnglish + powers[i];
            if (i > 0 && english.length > groupEnglish.length + powers[i].length) {
                english += " ";
            }
        }
    }

    return english;
}

function englishPower(digits) {
    let english = "";
    const hundreds = Math.floor(digits / 100);
    const remainder = digits % 100;
    const tensDigit = Math.floor(remainder / 10);
    const onesDigit = remainder % 10;

    if (hundreds !== 0) {
        english += ones[hundreds] + " hundred";
    }

    if (tensDigit === 1) {
        if (english) {
            english += " ";
        }
        english += teens[onesDigit];
        return english;
    } else {
        if (tensDigit !== 0) {
            if (english) {
                english += " ";
            }
            english += tens[tensDigit];
        }

        if (onesDigit !== 0) {
            if (english) {
                english += tensDigit > 0 ? '-' : ' ';
            }
            english += ones[onesDigit];
        }
    }

    return english;
}

module.exports = { inEnglish: inEnglish };